### PR TITLE
Feature/updated locale picker

### PR DESCRIPTION
**Pull request by khaykov**
This HW PR updates the look of the Application Locale picker. It's based on the Site Timezone picker.

You can filter by both label and localized label.

The original idea was to also provide locale suggestions, but because of the HW time constraints, I left it and some other potential improvements (getting rid of Pair and Triple) for the next HW.

Apologies for the size of the PR, but at the time feature flag seemed a bit of an overkill for the HW feature.

[Image from Gyazo](https://gyazo.com/ef74e9dab13e3dbd1fab40555d8e5b04)

To test:
- Navigate to App Settings, and open the Interface Language option.
- Confirm that the new locale picker works as expected:
      - You can filter locales by both their label and localized label (eg. typing "Dani" and "Dans" should produce list with "Danish/Dansk" locale)
      - Tapping on locale changes app language.
      - Current app locale is displayed a the top of the list. Tapping on it closes the list.

## Regression Notes
1. Potential unintended areas of impact
- Locale Picker is not working!

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Manual testing + VM tests.

3. What automated tests I added (or what prevented me from doing so)
- Hopefully all of them!

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding?.apply {
+            setupContentViews()
+            setupObservers()
+        }
+    }
+
+    private fun LocalePickerBottomSheetBinding.setupContentViews() {
+        val layoutManager = LinearLayoutManager(context)
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt]: LayoutManager is also set in layout xml, hence these two lines are not necessary here

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    override fun onBindViewHolder(holder: LocalePickerListViewHolder<*>, position: Int) {
+        val item = getItem(position)
+        when (holder) {
+            is LocalePickerListItemViewHolder -> holder.bind(item as LocaleRow)
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return getItem(position)!!.type.ordinal
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerAdapter.kt:25]: probably, not required given it's only one view type.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        btnLocaleSuggestion.setOnClickListener {
+            viewModel.onCurrentLocaleSelected()
+        }
+
+        dialog?.setOnShowListener { dialogInterface ->
+            val sheetDialog = dialogInterface as? BottomSheetDialog
+            bottomSheet = sheetDialog?.findViewById<View>(
+                    com.google.android.material.R.id.design_bottom_sheet
+            ) as? FrameLayout
+        }
```
Comment (ravishanker) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt:114]: Does it need to cater to landscape mode here?

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt (hunk 1 above)
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt]: Ah, good catch 👍

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerAdapter.kt (hunk 2 above)
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerAdapter.kt:25]: I'm planning to add a sub-header for preferred and all locales during next hack week, so I left this in :)

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt (hunk 3 above)
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/language/LocalePickerBottomSheet.kt:114]: I know that during timzeone picker work I mentioned that the bottom sheet is collapsed in landscape mode, but after some research, it seems like its intended behavior 🤷

**Comment by khaykov** [comment]
Thanks for the review, @ravishanker ! Hope I addressed all of your concerns :)

**Comment by AliSoftware** [comment]
I'm moving this to milestone 17.9 since I'm doing code freeze for 17.8 but this PR is still waiting for a review and approval and thus not ready to be merged before 17.8 code freeze; so it will make it to 17.9 instead.

Please ping me and your EW if you feel it's really important to make it to 17.8 and we could discussion it and consider retargetting it to the beta if it's really critical.

**Comment by khaykov** [comment]
This one totally fell off my radar - I realized I still have it open when going through my PR's. Since I'll be off for a couple of month I'm going to close this one, and reopen later after fixing conflicts.

**Comment by SiobhyB** [comment]
👋 @khaykov, I'm currently looking into an issue for groundskeeping that will involve adding additional languages to the Site Settings screen. I came across this PR as part of my research, as the addition of extra languages would make things hard to navigate with the current screen. The proposed changes in this PR seem like the perfect solution to that! 

I tested this out locally and, with just a few small tweaks, the code is still working. Would you be opposed to me re-basing this PR from `trunk` and re-opening it, with a view to getting it merged?

**Comment by khaykov** [comment]
Hey @SiobhyB 👋 Sure thing, I don't mind at all!

**Comment by SiobhyB** [comment]
Thanks so much @khaykov! It looks like there wasn't a way to update the base branch to `trunk` because `develop` has been deleted, the only option I was being provided with was to create a new PR, which I've gone ahead to do at https://github.com/wordpress-mobile/WordPress-Android/pull/16331 (your commits have been preserved). I'll work on it a little and set as ready for review again soon.

**Pull request closed without merging**